File: rust/src/taker.rs
//! Taker order signing types.

use alloy_primitives::Signature;
use alloy_signer::{Error as SignerError, SignerSync};
use alloy_signer_local::PrivateKeySigner;
use serde::{Deserialize, Serialize};
use smallvec::SmallVec;
use std::{error::Error, fmt};

use crate::types::{Address, Direction, MarketId, OrderType, MAX_RFQ_LEGS};

// =============================================================================
// ORDER LEG
// =============================================================================

/// Order leg for signing.
#[derive(Debug, Clone, Copy, Serialize, Deserialize)]
pub struct OrderLeg {
    /// Market identifier.
    pub market_id: MarketId,
    /// Direction (0=Up, 1=Down).
    ///
    /// For markets this encodes the boolean prediction:
    /// Up == price higher than strike, Down == price not higher than strike.
    pub direction: u8,
}

impl OrderLeg {
    /// Serialize leg to bytes for signing.
    pub fn to_bytes(&self) -> [u8; 9] {
        let mut bytes = [0u8; 9];
        bytes[..8].copy_from_slice(&self.market_id.as_u64().to_le_bytes());
        bytes[8] = self.direction;
        bytes
    }
}

// =============================================================================
// SIGNED ORDER
// =============================================================================

/// An order signed by the user's wallet.
///
/// This represents the user's commitment to a trade. The signature authorizes
/// the API to fill at or above `min_odds_bps`.
#[derive(Debug, Clone, Serialize, Deserialize)]
#[serde(bound = "")]
pub struct SignedOrder {
    // === Order Data ===
    /// User's wallet address (signer).
    pub user: Address,

    /// Wager amount in micro-dollars.
    pub wager_micros: u64,

    /// Minimum acceptable odds in basis points (`25_000 = 2.5x`).
    pub min_odds_bps: u32,

    /// Order legs.
    pub legs: Vec<OrderLeg>,

    /// Unique nonce to prevent replay.
    pub nonce: u64,

    /// Submission deadline in milliseconds since Unix epoch.
    pub expires_at_ms: u64,

    /// Order type: IOC or FOK.
    pub order_type: OrderType,

    /// If true, hide tier/address metadata from market makers.
    pub shield_on: bool,

    // === Signature ===
    /// ECDSA signature (65 bytes: r, s, v).
    #[serde(with = "signature_hex")]
    pub signature: [u8; 65],
}

mod signature_hex {
    use serde::{Deserialize, Deserializer, Serializer};

    pub fn serialize<S>(signature: &[u8; 65], serializer: S) -> Result<S::Ok, S::Error>
    where
        S: Serializer,
    {
        serializer.serialize_str(&hex::encode(signature))
    }

    pub fn deserialize<'de, D>(deserializer: D) -> Result<[u8; 65], D::Error>
    where
        D: Deserializer<'de>,
    {
        let mut value = String::deserialize(deserializer)?;
        if let Some(stripped) = value.strip_prefix("0x") {
            value = stripped.to_string();
        }

        let bytes = hex::decode(&value).map_err(serde::de::Error::custom)?;
        bytes
            .try_into()
            .map_err(|_| serde::de::Error::custom("expected 65-byte hex signature"))
    }
}

/// Invalid signed-order data.
#[derive(Debug, Clone, PartialEq, Eq)]
pub enum SignedOrderError {
    MissingLegs,
    TooManyLegs { max: usize, actual: usize },
    InvalidLegDirection { index: usize, direction: u8 },
}

impl fmt::Display for SignedOrderError {
    fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
        match self {
            Self::MissingLegs => write!(f, "order must include at least one leg"),
            Self::TooManyLegs { max, actual } => {
                write!(f, "order has {actual} legs, maximum is {max}")
            }
            Self::InvalidLegDirection { index, direction } => {
                write!(f, "invalid direction {direction} at legs[{index}]")
            }
        }
    }
}

impl Error for SignedOrderError {}

#[derive(Debug)]
pub enum TakerSignError {
    Order(SignedOrderError),
    Signer(SignerError),
}

impl fmt::Display for TakerSignError {
    fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
        match self {
            Self::Order(err) => write!(f, "{err}"),
            Self::Signer(err) => write!(f, "{err}"),
        }
    }
}

impl Error for TakerSignError {
    fn source(&self) -> Option<&(dyn Error + 'static)> {
        match self {
            Self::Order(err) => Some(err),
            Self::Signer(err) => Some(err),
        }
    }
}

impl From<SignedOrderError> for TakerSignError {
    fn from(err: SignedOrderError) -> Self {
        Self::Order(err)
    }
}

impl From<SignerError> for TakerSignError {
    fn from(err: SignerError) -> Self {
        Self::Signer(err)
    }
}

impl SignedOrder {
    /// Maximum number of legs in an order.
    pub const MAX_LEGS: usize = MAX_RFQ_LEGS;

    pub(crate) fn validate_legs(&self) -> Result<(), SignedOrderError> {
        if self.legs.is_empty() {
            return Err(SignedOrderError::MissingLegs);
        }
        if self.legs.len() > Self::MAX_LEGS {
            return Err(SignedOrderError::TooManyLegs {
                max: Self::MAX_LEGS,
                actual: self.legs.len(),
            });
        }
        for (index, leg) in self.legs.iter().enumerate() {
            if Direction::from_u8(leg.direction).is_none() {
                return Err(SignedOrderError::InvalidLegDirection {
                    index,
                    direction: leg.direction,
                });
            }
        }
        Ok(())
    }

    /// Try to serialize the funding-neutral order identity used for replay keying.
    ///
    /// This intentionally excludes the RFQ funding choice so attempts to reuse
    /// the same order with a different funding choice resolve to the same replay
    /// key and can be rejected as conflicting retries.
    ///
    /// Wire format:
    /// - `bytes[0..20]`: user address
    /// - `bytes[20..28]`: wager_micros (u64 LE)
    /// - `bytes[28..32]`: min_odds_bps (u32 LE)
    /// - `bytes[32..40]`: nonce (u64 LE)
    /// - `bytes[40..48]`: expires_at_ms (u64 LE)
    /// - `bytes[48]`: order_type (u8) - 1=IOC, 2=FOK
    /// - `bytes[49]`: shield_on (u8)
    /// - `bytes[50]`: leg_count (u8)
    /// - `bytes[51..]`: legs (9 bytes each)
    pub fn try_replay_bytes(&self) -> Result<SmallVec<[u8; 128]>, SignedOrderError> {
        self.try_order_bytes(None)
    }

    /// Serialize the funding-neutral order identity used for replay keying.
    ///
    /// Panics if the order is invalid. Use
    /// [`try_replay_bytes`](Self::try_replay_bytes) when handling externally
    /// constructed or mutated orders.
    #[inline]
    pub fn replay_bytes(&self) -> SmallVec<[u8; 128]> {
        self.try_replay_bytes().expect("invalid signed order")
    }

    /// Try to serialize order data and its funding choice to bytes for signing.
    ///
    /// Wallet signatures are EIP-191 signatures over these bytes, not over the
    /// JSON request payload. Use [`sign_order`] or [`signed_order`] to apply the
    /// same signing scheme as the server verifies.
    ///
    /// Wire format:
    /// - `bytes[0..20]`: user address
    /// - `bytes[20..28]`: wager_micros (u64 LE)
    /// - `bytes[28..32]`: min_odds_bps (u32 LE)
    /// - `bytes[32..40]`: nonce (u64 LE)
    /// - `bytes[40..48]`: expires_at_ms (u64 LE)
    /// - `bytes[48]`: order_type (u8) - 1=IOC, 2=FOK
    /// - `bytes[49]`: shield_on (u8)
    /// - `bytes[50]`: use_app_tokens (u8)
    /// - `bytes[51]`: leg_count (u8)
    /// - `bytes[52..]`: legs (9 bytes each)
    ///
    /// SECURITY: `order_type` is included in the signature to prevent replay
    /// attacks where an attacker could submit a signed FOK order as IOC.
    pub fn try_signing_bytes(
        &self,
        use_app_tokens: bool,
    ) -> Result<SmallVec<[u8; 128]>, SignedOrderError> {
        self.try_order_bytes(Some(use_app_tokens))
    }

    fn try_order_bytes(
        &self,
        use_app_tokens: Option<bool>,
    ) -> Result<SmallVec<[u8; 128]>, SignedOrderError> {
        self.validate_legs()?;
        let mut bytes = SmallVec::with_capacity(
            51 + usize::from(use_app_tokens.is_some()) + self.legs.len() * 9,
        );

        // User address (20 bytes)
        bytes.extend_from_slice(self.user.as_slice());

        // Wager (8 bytes LE)
        bytes.extend_from_slice(&self.wager_micros.to_le_bytes());

        // Min odds (4 bytes LE)
        bytes.extend_from_slice(&self.min_odds_bps.to_le_bytes());

        // Nonce (8 bytes LE)
        bytes.extend_from_slice(&self.nonce.to_le_bytes());

        // Submission expiry (8 bytes LE)
        bytes.extend_from_slice(&self.expires_at_ms.to_le_bytes());

        // Order type (1 byte)
        bytes.push(self.order_type as u8);

        // Shield flag (1 byte)
        bytes.push(u8::from(self.shield_on));

        // App-token funding flag (1 byte, signing payload only)
        if let Some(use_app_tokens) = use_app_tokens {
            bytes.push(u8::from(use_app_tokens));
        }

        // Leg count (1 byte)
        bytes.push(self.legs.len() as u8);

        // Legs (9 bytes each)
        for leg in &self.legs {
            bytes.extend_from_slice(&leg.to_bytes());
        }

        Ok(bytes)
    }

    /// Serialize order data to bytes for signing.
    ///
    /// Panics if the order is invalid. Use [`try_signing_bytes`](Self::try_signing_bytes)
    /// when handling externally constructed or mutated orders.
    #[inline]
    pub fn signing_bytes(&self, use_app_tokens: bool) -> SmallVec<[u8; 128]> {
        self.try_signing_bytes(use_app_tokens)
            .expect("invalid signed order")
    }

    /// Verify this order's signature and funding choice against `self.user`.
    pub fn verify_signature(&self, use_app_tokens: bool) -> bool {
        let Ok(signing_bytes) = self.try_signing_bytes(use_app_tokens) else {
            return false;
        };
        Signature::from_raw_array(&self.signature)
            .ok()
            .map(|signature| signature.normalized_s())
            .and_then(|signature| signature.recover_address_from_msg(&signing_bytes).ok())
            .is_some_and(|recovered| recovered == self.user)
    }
}

/// Sign a taker order with EIP-191 over [`SignedOrder::try_signing_bytes`].
pub fn sign_order(
    order: &mut SignedOrder,
    use_app_tokens: bool,
    signing_key: &PrivateKeySigner,
) -> Result<(), TakerSignError> {
    let signing_bytes = order.try_signing_bytes(use_app_tokens)?;
    order.signature = signing_key.sign_message_sync(&signing_bytes)?.into();
    Ok(())
}

/// Return a signed taker order, preserving all order fields and replacing the signature.
pub fn signed_order(
    mut order: SignedOrder,
    use_app_tokens: bool,
    signing_key: &PrivateKeySigner,
) -> Result<SignedOrder, TakerSignError> {
    sign_order(&mut order, use_app_tokens, signing_key)?;
    Ok(order)
}

File: typescript/src/taker.ts
import { recoverMessageAddress } from "viem";
import { privateKeyToAccount } from "viem/accounts";

import {
  bytesFrom,
  bytesToHex,
  checkU32,
  checkU64,
  concatBytes,
  encodeBase64,
  u32LeBytes,
  u64LeBytes,
} from "./bytes.js";
import { MAX_RFQ_LEGS } from "./rfq.js";
import { Address, Direction, MarketId, OrderType, u64SerdeValue } from "./types.js";

export class SignedOrderError extends Error {
  constructor(message: string) {
    super(message);
    this.name = "SignedOrderError";
  }
}

export class TakerSignError extends Error {
  constructor(message: string, readonly cause?: unknown) {
    super(message);
    this.name = "TakerSignError";
  }
}

export interface OrderLegInput {
  marketId?: MarketId | number | bigint | string;
  market_id?: MarketId | number | bigint | string;
  direction: number;
}

export class OrderLeg {
  readonly marketId: MarketId;
  readonly direction: number;

  constructor(input: OrderLegInput) {
    const marketId = input.marketId ?? input.market_id;
    if (marketId === undefined) {
      throw new Error("market_id is required");
    }
    this.marketId = marketId instanceof MarketId ? marketId : MarketId.new(marketId);
    if (!Number.isInteger(input.direction) || input.direction < 0 || input.direction > 0xff) {
      throw new RangeError("direction must fit in u8");
    }
    this.direction = input.direction;
  }

  toBytes(): Uint8Array {
    return concatBytes(u64LeBytes(this.marketId.asU64()), Uint8Array.of(this.direction));
  }

  serdeValue(): { market_id: number | bigint; direction: number } {
    return { market_id: this.marketId.serdeValue(), direction: this.direction };
  }
}

export interface SignedOrderInput {
  user: Address | string | Uint8Array;
  wagerMicros?: number | bigint | string;
  wager_micros?: number | bigint | string;
  /** Minimum acceptable odds in basis points (`25_000 = 2.5x`). */
  minOddsBps: number;
  legs: Array<OrderLeg | OrderLegInput>;
  nonce: number | bigint | string;
  expiresAtMs?: number | bigint | string;
  expires_at_ms?: number | bigint | string;
  orderType?: OrderType | number;
  order_type?: OrderType | number;
  shieldOn?: boolean;
  shield_on?: boolean;
  signature?: Uint8Array | string | number[];
}

export class SignedOrder {
  static readonly MAX_LEGS = MAX_RFQ_LEGS;

  readonly user: Address;
  readonly wagerMicros: number | bigint | string;
  /** Signed basis-point value; HTTP `SignedOrderJson.min_odds` uses decimal odds. */
  readonly minOddsBps: number;
  readonly legs: OrderLeg[];
  readonly nonce: number | bigint | string;
  readonly expiresAtMs: number | bigint | string;
  readonly orderType: OrderType | number;
  readonly shieldOn: boolean;
  readonly signature: Uint8Array;

  constructor(input: SignedOrderInput) {
    this.user = Address.fromEvm(input.user);
    this.wagerMicros = input.wagerMicros ?? input.wager_micros ?? missing("wager_micros");
    this.minOddsBps = input.minOddsBps ?? missing("min_odds_bps");
    this.legs = input.legs.map((leg) => (leg instanceof OrderLeg ? leg : new OrderLeg(leg)));
    this.nonce = input.nonce;
    this.expiresAtMs = input.expiresAtMs ?? input.expires_at_ms ?? missing("expires_at_ms");
    this.orderType = input.orderType ?? input.order_type ?? missing("order_type");
    const shieldOn = input.shieldOn !== undefined ? input.shieldOn : input.shield_on;
    this.shieldOn = shieldOn === undefined ? false : shieldOn;
    this.signature =
      input.signature === undefined ? new Uint8Array(65) : bytesFrom(input.signature, 65, "signature");
  }

  validate(): void {
    if (typeof this.shieldOn !== "boolean") {
      throw new SignedOrderError("shield_on must be bool");
    }
    if (this.legs.length === 0) {
      throw new SignedOrderError("order must include at least one leg");
    }
    if (this.legs.length > SignedOrder.MAX_LEGS) {
      throw new SignedOrderError("order legs exceed MAX_LEGS");
    }
    this.legs.forEach((leg, index) => {
      if (Direction.fromU8(leg.direction) === undefined) {
        throw new SignedOrderError(`invalid direction ${leg.direction} at legs[${index}]`);
      }
    });
    if (this.orderType !== OrderType.IOC && this.orderType !== OrderType.FOK) {
      throw new SignedOrderError(`invalid order type: ${this.orderType}`);
    }
    checkU64(this.wagerMicros, "wager_micros");
    checkU32(this.minOddsBps, "min_odds_bps");
    checkU64(this.nonce, "nonce");
    checkU64(this.expiresAtMs, "expires_at_ms");
  }

  signingBytes(useAppTokens: boolean): Uint8Array {
    this.validate();
    if (typeof useAppTokens !== "boolean") {
      throw new SignedOrderError("use_app_tokens must be bool");
    }
    const parts = [
      this.user.asSlice(),
      u64LeBytes(this.wagerMicros),
      u32LeBytes(this.minOddsBps),
      u64LeBytes(this.nonce),
      u64LeBytes(this.expiresAtMs),
      Uint8Array.of(
        Number(this.orderType),
        this.shieldOn ? 1 : 0,
        useAppTokens ? 1 : 0,
        this.legs.length,
      ),
      ...this.legs.map((leg) => leg.toBytes()),
    ];
    return concatBytes(...parts);
  }

  /** Verify this order's EIP-191 signature and funding choice against `this.user`. */
  async verifySignature(useAppTokens: boolean): Promise<boolean> {
    try {
      const recovered = await recoverMessageAddress({
        message: { raw: bytesToHex(this.signingBytes(useAppTokens), true) as `0x${string}` },
        signature: bytesToHex(this.signature, true) as `0x${string}`,
      });
      return Address.fromEvm(recovered).equals(this.user);
    } catch {
      return false;
    }
  }

  serdeValue(): {
    user: string;
    wager_micros: number | string;
    min_odds: number;
    legs: Array<{ market_id: number | bigint; direction: "up" | "down" }>;
    nonce: number | string;
    expires_at_ms: number | string;
    order_type: number;
    shield_on: boolean;
    signature: string;
  } {
    this.validate();
    return {
      user: this.user.toChecksum(),
      wager_micros: u64SerdeValue(this.wagerMicros),
      min_odds: this.minOddsBps / 10_000,
      legs: this.legs.map((leg) => ({
        market_id: leg.marketId.serdeValue(),
        direction: serdeDirection(leg.direction),
      })),
      nonce: u64SerdeValue(this.nonce),
      expires_at_ms: u64SerdeValue(this.expiresAtMs),
      order_type: serdeOrderType(this.orderType),
      shield_on: this.shieldOn,
      signature: encodeBase64(this.signature),
    };
  }
}

export async function signOrder(
  order: SignedOrder,
  useAppTokens: boolean,
  signingKey: `0x${string}`,
): Promise<SignedOrder> {
  const account = privateKeyToAccount(signingKey);
  const signature = await account.signMessage({
    message: { raw: bytesToHex(order.signingBytes(useAppTokens), true) as `0x${string}` },
  });
  return new SignedOrder({
    user: order.user,
    wagerMicros: order.wagerMicros,
    minOddsBps: order.minOddsBps,
    legs: order.legs,
    nonce: order.nonce,
    expiresAtMs: order.expiresAtMs,
    orderType: order.orderType,
    shieldOn: order.shieldOn,
    signature: bytesFrom(signature, 65, "signature"),
  });
}

export async function signedOrder(
  order: SignedOrder,
  useAppTokens: boolean,
  signingKey: `0x${string}`,
): Promise<SignedOrder> {
  return signOrder(order, useAppTokens, signingKey);
}

function missing(name: string): never {
  throw new Error(`${name} is required`);
}

function serdeDirection(direction: number): "up" | "down" {
  const parsed = Direction.fromU8(direction);
  if (parsed === Direction.Up) {
    return "up";
  }
  if (parsed === Direction.Down) {
    return "down";
  }
  throw new RangeError("invalid direction");
}

function serdeOrderType(orderType: OrderType | number): number {
  const raw = Number(orderType);
  if (OrderType.fromU8(raw) === undefined) {
    throw new RangeError("invalid order type");
  }
  return raw;
}
